src/platform/update/new_release_monitor.cpp

```diff
@@ -0,0 +1,155 @@
+
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QThread>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+namespace
+{
+constexpr auto timeout = std::chrono::minutes(1);
+constexpr auto json_tag_name = "tag_name";
+constexpr auto json_html_url = "html_url";
+
+QJsonObject parse_manifest(const QByteArray& json)
+{
+    QJsonParseError parse_error;
+    const auto doc = QJsonDocument::fromJson(json, &parse_error);
+    if (doc.isNull())
+        throw std::runtime_error(parse_error.errorString().toStdString());
+
+    if (!doc.isObject())
+        throw std::runtime_error("invalid JSON object");
+
+    return doc.object();
+}
+} // namespace
+
+class mp::LatestReleaseChecker : public QThread
+{
+    Q_OBJECT
+public:
+    LatestReleaseChecker(QString update_url) : update_url(update_url)
+    {
+    }
+
+    void run() override
+    {
+        QUrl url(update_url);
+        try
+        {
+            mp::URLDownloader downloader(::timeout);
+            QByteArray json = downloader.download(url);
+            const auto manifest = ::parse_manifest(json);
+            if (!manifest.contains(::json_tag_name) || !manifest.contains(::json_html_url))
+                throw std::runtime_error("Github JSON missing required fields");
+
+            mp::NewReleaseInfo release;
+            release.version = manifest[::json_tag_name].toString();
+            if (release.version[0] == 'v')
+                release.version.remove(0, 1);
+
+            release.url = manifest[::json_html_url].toString();
+
+            mpl::log(mpl::Level::debug, "update",
+                     fmt::format("Latest Multipass release available is version {}", qPrintable(release.version)));
+
+            emit latest_release_found(release);
+        }
+        catch (const mp::DownloadException& e)
+        {
+            mpl::log(mpl::Level::info, "update", fmt::format("Failed to fetch update info: {}", qPrintable(e.what())));
+        }
+        catch (const std::runtime_error& e)
+        {
+            mpl::log(mpl::Level::info, "update", fmt::format("Failed to parse update info: {}", qPrintable(e.what())));
+        }
+    }
+signals:
+    void latest_release_found(const multipass::NewReleaseInfo& release);
+
+private:
+    const QString update_url;
+};
+
+mp::NewReleaseMonitor::NewReleaseMonitor(const QString& current_version,
+                                         std::chrono::steady_clock::duration refresh_rate, const QString& update_url)
+    : current_version(current_version), update_url(update_url)
+{
+    qRegisterMetaType<multipass::NewReleaseInfo>(); // necessary to allow custom type be passed in signal/slot
+
+    check_for_new_release();
+
+    refresh_timer.setInterval(std::chrono::duration_cast<std::chrono::milliseconds>(refresh_rate));
+    connect(&refresh_timer, &QTimer::timeout, this, &mp::NewReleaseMonitor::check_for_new_release);
+    refresh_timer.start();
+}
+
+mp::NewReleaseMonitor::~NewReleaseMonitor() = default;
```

Comment:
Yep

Reply:
done

Reply:
Not done after all, this is required so that the dtor of `NewReleaseMonitor` is only defined after `LatestReleaseChecker`. The latter is only fwd-declared in the header, but it is composed with a `unique_ptr`, whose dtor requires the definition of the class. Since the `unique_ptr`'s dtor is called by the generated dtor, it can only be generated after the type of the pointee is known.